File: DistributeCoins.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 //Time - O(n)
 //sapce - O(logn)
class Solution {
    int moves =0;
    public int distributeCoins(TreeNode root) {
        if(root == null) return 0;
        dfs(root);
        return moves;
    }
    int dfs(TreeNode root){
        if(root == null) return 0;
        int left = dfs(root.left);
        int right = dfs(root.right);
        int extra = left+right+root.val-1;
        moves+= Math.abs(extra);
        return extra;
    }
}

File: distributeCoins.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int moves;
    int helper (TreeNode* root) {
        if (root==NULL) return 0;
        int extra=(root->val)-1;
        
        extra+=helper(root->left);
        extra+=helper(root->right);
        moves+=abs(extra);
        
        return extra;
    }
    int distributeCoins(TreeNode* root) {
        moves=0;
        helper(root);
        return moves;
    }
};